python/Ganga/GPIDev/Base/Proxy.py

```diff
@@ -561,16 +570,25 @@ def GPIProxyObjectFactory(_obj):
         from Ganga.Core.exceptions import GangaException
         raise GangaException("%s is NOT a Proxyable object" % type(_obj))
 
-    if hasattr(_obj, proxyObject):
-        return getattr(_obj, proxyObject)
+    ## This is defined within Objects.py, we could probably store this elsehere
+    ## (We probably do) but as this is guaranteed to be accessible for GangaObjects I will use this
+    this_class = getattr(stripProxy(_obj).__class__, proxyClass)
 
-    if not hasattr(_obj.__class__, proxyClass):
-        from Ganga.Core.exceptions import GangaException
-        raise GangaException("%s is NOT a Proxyable object" % type(_obj))
+    ## Would LOVE to avoid having to perform all of these checks to get or wrap a class in a proxy here but
+    ## multiple different BAD states are currently passed to this function
+    ## In future I would like to see this fixed elsewhere and reduce the logic in this section of code
+    ## As of Ganga 6.1.15 this is needed to put objects into a semi-coherent state
 
-    proxy_class = getattr(_obj.__class__, proxyClass)
+    if isType(_obj, this_class):
+        if isProxy(_obj):
+            return _obj
+        elif not isProxy(_obj):
+            pass
+    else:
+        pass
 
-    return proxy_class(_existing=_obj)
+    proxy_class = this_class(_proxy_impl_obj_to_wrap=_obj)
+    return proxy_class
```

Comment:
You're right that at present `_proxyClass` is always present on a `GangaObject`. In future I was planning on not always making it so, such that we could have some `GangaObject`s as non-proxiable (such as `GangaObject` itself, the adapters etc.). In this world we would have to be explicit about which classes we export.

I would want to keep any changes to the proxy system in a separate RP so that we can discuss them in their own context and not mixed up with speeding up running `jobs`.